(CLOUD-295) Add RDS resource support

garethr opened the pull request:
This takes on the work from @petems in https://github.com/puppetlabs/puppetlabs-aws/pull/161 and:

* Fixes the failing tests
* Adds lots more tests
* Adds types checking of parameters and properties
* Removes a number of intermediary commits
* Standardises formatting of property descriptions in the type
* Expand the assertions in the acceptance tests
* Add documentation for the new types and properties to the README

petems commented:
Awesome! :+1:

jbondpdx commented on README.md:
```diff
@@ -775,11 +778,127 @@ routes => [
-#### Type: route53*
+#### Type: rds_db_parameter_group
+
+Note that this type is currently only available to list via puppet
```
I am not totally sure what this means. I think it is something like: This type can only be listed via `puppet resource

jbondpdx commented on README.md:
```diff
@@ -775,11 +778,127 @@ routes => [
+*Required* The region in which to launch the parameter group. For valid values, see [AWS Regions](http://docs.aws.amazon.com/general/latest/gr/rande.html#ec2_region).
+
+#####`description`
+A description of the parameter group.
```
All of these new parameters should say what the acceptable or valid values are. Something like "Accepts a string." or "Valid options: true or false." should be fine (assuming that's correct, of course!)

jbondpdx commented on README.md:
```diff
@@ -775,11 +778,127 @@ routes => [
+#####`description`
+A description of the parameter group.
+
+#####`family`
```
The name of the database family with which the parameter group is compatible; for instance, 'mysql5.1'.

jbondpdx commented on README.md:
```diff
@@ -775,11 +778,127 @@ routes => [
+*Required* The region in which to launch the parameter group. For valid values, see [AWS Regions](http://docs.aws.amazon.com/general/latest/gr/rande.html#ec2_region).
+
+#####`owner_id`
+The internal AWS id of the owner of the security group. Read-only
```
I take this to mean read-only ever, not read-only after creation? If so, looks fine, but needs a period after Read-only. Same for similar params below.

jbondpdx commented on README.md:
```diff
@@ -775,11 +778,127 @@ routes => [
+#### Type: rds_instance
+
+#####`name`
+*Required* The name of the RDS Instance
```
needs a period at the end.

jbondpdx commented on README.md:
```diff
@@ -775,11 +778,127 @@ routes => [
+#####`name`
+*Required* The name of the RDS Instance
+
+#####`db_name`
```
missing a space between 'this' and 'is'

jbondpdx commented on README.md:
```diff
@@ -775,11 +778,127 @@ routes => [
+using the `rds-describe-db-engine-versions` command from the AWS CLI.
+
+#####`allocated_storage`
+The size of the database in gigabytes. Note that minimum size constrains
```
constraints is missing the 't', and there should be a comma after 'exist'.

jbondpdx commented on README.md:
```diff
@@ -775,11 +778,127 @@ routes => [
+
+#####`multi_az`
+Boolean. Required if you intend to run the instance across multiple
+availabililty zones.
```
availability (typo)

jbondpdx commented on examples/postgres-rds-example/README.md:
```diff
@@ -0,0 +1,53 @@
+
+## How
+
+This example creates a security group to allow Postgres RDS instance, then creates an RDS instance with that security group assigned.
```
missing article: "...to allow a Postrgres RDS instance..."

jbondpdx commented on examples/postgres-rds-example/README.md:
```diff
@@ -0,0 +1,53 @@
+
+    puppet apply rds_security.pp
+
+Unfortunatly, it's not possible to assign the EC2 group and the allowed IP's to the `db_securitygroup` through the API, so you have to do this manually though the console for now:
```
typo: Unfortunately

Also, IPs should not have an apostrophe.

jbondpdx commented on examples/postgres-rds-example/README.md:
````diff
@@ -0,0 +1,53 @@
+
+It should return something like this:
+
+```puppet
````
prefer ~~~ to ``` for code fences because of compatibility with our main docs site.

jbondpdx commented:
@garethr and @petems , I've made comments on the readmes; however,  an alternative would be to just merge the docs as they are, and then I can take an editorial pass at them after that. I'd be happy to do do that and PR docs-only changes.

garethr commented:
@jbondpdx I've updated the PR with all the suggested changes, and I've rebased with master to include the other README updates.

jbondpdx commented:
Awesome, thank you, @garethr . +1

DavidS commented on lib/puppet/provider/rds_instance/v2.rb:
```diff
@@ -0,0 +1,115 @@
+  def destroy
+    Puppet.info("Deleting database #{name} in region #{resource[:region]}")
+    rds = rds_client(resource[:region])
+    Puppet.info("Skip Final Snapshot: #{resource[:skip_final_snapshot]}")
```
It would be nice to have a proper message saying what will or will not be done here, instead of just dumping the value.

garethr replied:
Agreed, fixed.

DavidS commented on lib/puppet/provider/rds_instance/v2.rb:
```diff
@@ -0,0 +1,115 @@
+      skip_final_snapshot: resource[:skip_final_snapshot].to_s,
+      final_db_snapshot_identifier: resource[:final_db_snapshot_identifier],
+    }
+    rds.delete_db_instance(config)
```
I'm not familiar with aws api, but this looks as if I can set skip_final_snapshot to any value I like at destroy time
why is it marked as readonly then ?

garethr replied:
Good catch. On further investigation here these should be params not properties, they aren't set at creation and they aren't readable back from AWS. They only affect things at deletion time. I've pushed the changes and related test fixes. Cheers.

DavidS commented on tests/create.pp:
```diff
@@ -89,4 +89,4 @@
     instance_port      => 80,
   }],
-}
+}
```
??

garethr replied:
No idea either, but I've killed it with fire.

garethr commented:
@DavidS I've updated this PR with suggestions and rebased against master. It would be useful to review sooner rather than later in order to make sure @zreichert has time to review in this sprint once it's been through CI. Cheers

DavidS commented on lib/puppet/provider/rds_db_parameter_group/v2.rb:
```diff
@@ -0,0 +1,37 @@
+    [:present, :creating, :available].include? @property_hash[:ensure]
+  end
+
+end
```
When trying to create at rds_db_parameter_group, puppet errors badly: 
```
Error: Could not set 'present' on ensure: undefined method `create' for Rds_db_parameter_group[name=test2]:Puppet::Type::Rds_db_parameter_group
```

It'd be great if we could add create and destroy methods that fail with useful error messages.

(also, missing newline at EOF)

garethr replied:
Actually, why is that even ensurable? I've removed that, much more sensible:

```
puppet resource rds_db_parameter_group garethr ensure=present region=sa-east-1
Error: Could not run: Invalid parameter ensure
```

DavidS commented on lib/puppet/type/rds_instance.rb:
```diff
@@ -0,0 +1,202 @@
+Must contain 1 to 63 alphanumeric characters
+Must begin with a letter or an underscore. Subsequent characters can be letters, underscores, or digits (0-9).
+Cannot be a word reserved by the specified database engine
+Oracle
```
Missing newline before this

garethr replied:
Added

DavidS commented on lib/puppet/type/rds_instance.rb:
```diff
@@ -0,0 +1,202 @@
+Constraints:
+
+Cannot be longer than 8 characters
+SQL Server
```
Missing newline before this

garethr replied:
Added

DavidS commented:
I see that rds_db_parameter_group is currently quite limited. It'd be great to have at least one acceptance test for it though.

DavidS commented:
* Listing a `rds_db_securitygroup` with puppet resource doesn't show the description.
  * Running puppet resource with ensure=absent, doesn't remove it.

```
vagrant@localhost:/vagrant/git/puppetlabs-aws$ AWS_REGION=eu-west-1 rbenv exec bundle exec puppet resource --libdir=./lib --debug rds_db_securitygroup test
Debug: Runtime environment: puppet_version=3.7.5, ruby_version=1.9.3, run_mode=user, default_encoding=UTF-8
Info: Checking if DB Security Group test exists
rds_db_securitygroup { 'test':
  ensure   => 'present',
  owner_id => '482693910459',
  region   => 'eu-west-1',
}
vagrant@localhost:/vagrant/git/puppetlabs-aws$ AWS_REGION=eu-west-1 rbenv exec bundle exec puppet resource --libdir=./lib --debug rds_db_securitygroup test ensure=absent
Debug: Runtime environment: puppet_version=3.7.5, ruby_version=1.9.3, run_mode=user, default_encoding=UTF-8
Debug: Loaded state in 0.00 seconds
Debug: Prefetching v2 resources for rds_db_securitygroup
Info: Checking if DB Security Group test exists
Debug: /Rds_db_securitygroup[test]: Nothing to manage: no ensure and the resource doesn't exist
Debug: Finishing transaction 16506260
Debug: Storing state
Debug: Stored state in 0.01 seconds
Info: Checking if DB Security Group test exists
rds_db_securitygroup { 'test':
  ensure => 'absent',
}
vagrant@localhost:/vagrant/git/puppetlabs-aws$ AWS_REGION=eu-west-1 rbenv exec bundle exec puppet resource --libdir=./lib --debug rds_db_securitygroup test
Debug: Runtime environment: puppet_version=3.7.5, ruby_version=1.9.3, run_mode=user, default_encoding=UTF-8
Info: Checking if DB Security Group test exists
rds_db_securitygroup { 'test':
  ensure   => 'present',
  owner_id => '482693910459',
  region   => 'eu-west-1',
}
vagrant@localhost:/vagrant/git/puppetlabs-aws$
```

garethr commented:
Fixed 

* puppet resource not showing the description for db_securitygroups
* removed ensure from the db_parameter_group

The issue you're seeing with ensure=absent is that you always have to specify the region in the params, not just the ENV variable. This is on purpose, but I think we should change it so what you expected to work just works. See https://github.com/puppetlabs/puppetlabs-aws/pull/117, although that work needs prioritising.
